ENH: Lagrangian interpolation for UT1-UTC in IERS data

MaanasArora opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This PR implements cubic Lagrangian interpolation (instead of linear interpolation) for only UT1-UTC in IERS data as suggested in #1803. Currently does not correct for luni-solar tides (step 2 and 4).

I'm new to this project, so hope this aligns with the feature request! There was some confusion around verifying the procedure in the linked issue it seems, so open to discussing further--thank you for reviewing!

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #1803.

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

MaanasArora committed:
ENH: Simple lagrange interpolation for UT1-UTC in IERS

MaanasArora committed:
BUG: Fix issues with UT1-UTC leap second jump handling

MaanasArora committed:
MAINT: Remove unused param

pllim commented:
CI is failing so I think this is too risky to go into v7.1 given how close we are to feature freeze. I have turned this into draft for and milestoned it for v7.2. Thank you for your interest and understanding.

MaanasArora commented:
No worries, thanks for looking at this.

mhvk commented on astropy/utils/iers/iers.py:
```diff
@@ -442,6 +442,36 @@ class because it has different requirements.
+            for j in range(len(x_vals)):
+                if i != j:
+                    diff = x_vals[i] - x_vals[j]
+                    diff = np.where(diff == 0, 1e-10, diff)
```
Within the IERS table, it is guaranteed that the `x` values are never the same, so this shouldn't be necessary.

mhvk commented on astropy/utils/iers/iers.py:
```diff
@@ -442,6 +442,36 @@ class because it has different requirements.
+
+    def _ut1_utc_interp(self, i0, i1, mjd):
+        indices = np.array([i0 - 1, i0, i1, i1 + 1])
+        indices = np.clip(indices, 0, len(self) - 1)
```
I think you'll need to set up the clipping of `i1` on l.462 (pre-change) in `interpolate` differently instead of redoing the clipping here. Note that if you look at https://hpiers.obspm.fr/iers/models/interp.f, it is not the indices that are clipped - rather it is made sure that there are always three values.

mhvk commented on astropy/utils/iers/iers.py:
```diff
@@ -442,6 +442,36 @@ class because it has different requirements.
+        vals[1:] += leap_diff
+
+        val = self._lagrange_interp(mjds, vals, mjd)
+        val -= leap_diff[0]
```
I think in the general case you cannot be sure index 0 is the right one to subtract. Which one it is depends on how the 3 surrounding points are selected, so this likely is an array.

mhvk reviewed (commented):
Thanks for the PR! Some smaller comments are in-line. 

Bigger ones: As @pllim notes, first thing would be to make sure that the code passes tests. Note that the failures may well be because the answer actually changes a little. But we should not simply adjust test, but rather expand them for two choices, of either linear interpolation ignoring tides, or quadratic including them (see also below).

More generally: if we go this route, we will also want to interpolate other columns with higher order, so it may make more sense to try to have one interpolation routine rather than special-case `UT1-UTC`. The only reason `UT1_UTC` is different from `X` and `Y` is because of the leap seconds, so perhaps one can find some generic way to deal with that, then interpolate either linearly or quadratically, and then take into account the leap second again.

Another general note: I think that without the tides there is not so much point in doing the quadratic interpolation. Also, eventually, since the tides is a relatively heavy computation, we probably want a `ScienceState` configuration to decide how to do it.

Finally, just a note: we are very close to feature freeze, so I won't have time to help much the coming weeks. I do really like the idea of closing that very old issue, though!

pllim commented:
> tides is a relatively heavy computation

Probably also need benchmarking then?

MaanasArora commented:
Thank you @mhvk ; these all make sense! I'm starting by generalizing the routine to allow (optionally) interpolation for any column / order, which should clear a lot of the issues, next I'll add the tides logic.

MaanasArora committed:
ENH: Add flexible interpolation option to IERS interpolate and improve leap second handling for UT1-UTC

MaanasArora commented:
Thanks again @mhvk, the pointers were really helpful!

I have provisioned for allowing Lagrange interpolation over all columns, and the order can be changed from the default (cubic), but it's the same for all columns; I suppose we would want different orders for different columns (higher for X and Y perhaps), but what would they be (at least the defaults)?

I think I've addressed the minor points--the leap second handling and index clipping is now more sophisticated, and depends on the order.

One can select linear interpolation and it is enabled by default, so the tests passed as usual. Thinking of adding new tests for the other interpolation methods but perhaps we can clarify the details and add the tides first. If this looks fine I can go ahead and add the tides!

MaanasArora commented:
The failing test is for the changelog -- not sure if anything needs to be done there!

MaanasArora committed:
MAINT: Improve argument name for IERS interpolation order

MaanasArora committed:
ENH: Add interpolation parameters to interpolate functions in IERS

MaanasArora committed:
BUG: Fix incorrect leap second correction logic in IERS interpolate

MaanasArora committed:
TST: Add parametrization for interpolation including Lagrange to IERS tests

MaanasArora commented:
I've made some improvements, including a fix to the leap second offset index, and added some parametrized coverage for interpolation to `test_iers.py`. Seems the tests pass now!

MaanasArora committed:
Merge branch 'main' into enh/lagrange-interp-ut1-utc

MaanasArora committed:
Merge branch 'main' into enh/lagrange-interp-ut1-utc

MaanasArora committed:
ENH: Implement ray-luni-solar tides correction in IERS class for UT1-UTC transformations

MaanasArora commented:
Hi @mhvk, I've added adjustment for the luni-solar tides! I'll double check again the constants as there were so many, and then maybe should add an argument? 

Would appreciate a look at this. Thank you!

MaanasArora committed:
BUG: Remove unnecessary whitespace

MaanasArora committed:
DOC: Add changelog entry

MaanasArora committed:
Merge branch 'main' into enh/lagrange-interp-ut1-utc

MaanasArora committed:
Merge branch 'main' into enh/lagrange-interp-ut1-utc

MaanasArora committed:
Merge branch 'main' into enh/lagrange-interp-ut1-utc

MaanasArora committed:
Merge branch 'main' into enh/lagrange-interp-ut1-utc

MaanasArora committed:
Merge branch 'main' into enh/lagrange-interp-ut1-utc

mhvk commented:
@MaanasArora - I'm very sorry that this completely slipped through the cracks... I only got reminded because the bot is trying to close this PR (I've removed the `close?` label for now).

I think the code itself is good, but I wonder about how best to expose it to the user. Generally, someone does not use this code directly, but it gets called when transforming a time or coordinate. Passing up arguments seems a non-starter; instead, I think these belong as configuration items in `iers.conf` -- see the `Conf` class. I think it is not super helpful to make this incredibly flexible. E.g., I could imagine having an `iers_interpolation_scheme` configuration option that can take values of
- "linear" or "astropy03": to get the old scheme
- "tides" or "astropy81" ("astropy90"?): new lagrange + tides

I guess in principle one could also have "lagrange" which omits the tides, but it is not obvious to me that is useful: is it actually closer to the truth than linear? It would need testing. I do not think we should expose things like `order`

On "the truth", we do need something to compare with! I guess easiest would be to run `interp.f`. (Indeed, I now wonder if we cannot just use numpy's `f2py` to translate that to a python module, even if just for testing.)

mhvk commented:
@MaanasArora - turned out `interp.f` was actually fairly easily compilable with `f2py` (see below). We can use this to verify results (I even wondered about just using it, but do not know we can start relying on a fortran compiler being present...)

Interpolation of IERS tables
----------------------------

The file ``interp.f`` was downloaded from
https://hpiers.obspm.fr/iers/models/interp.f, untabified in emacs, and
the following patches applied:
```
29a30
>       use, intrinsic :: ieee_arithmetic, only: IEEE_Value, IEEE_QUIET_NAN
31c32
<       integer i,n
---
>       integer n
35a37,42
>       IF (rjd_int .lt. RJD(1) .or. rjd_int .gt. RJD(n)) THEN
>          x_int = IEEE_VALUE(x_int, IEEE_QUIET_NAN)
>          y_int = x_int
>          ut1_int = x_int
>          RETURN
>       END IF
86a94
>       k = 0
90a99
>       if ( k .eq. 0 ) return
```

A draft signature file was made by
```
f2py interp.f -m interp_mod -h interp.pyf
```
and then edited to make clear what variables are input and output
ones, leading to:
```
!    -*- f90 -*-
! Note: the context of this file is case sensitive.

python module interp_mod ! in
    interface  ! in :interp
        subroutine interp(rjd,x,y,ut1,n,rjd_int,x_int,y_int,ut1_int) ! in :interp:interp.f
            double precision dimension(n), intent(in) :: rjd
            double precision dimension(n), intent(in) :: x
            double precision dimension(n), intent(in) :: y
            double precision dimension(n), intent(in) :: ut1
            integer, intent(hide), depend(rjd) :: n=shape(rjd, 0)
            double precision :: rjd_int
            double precision, intent(out) :: x_int
            double precision, intent(out) :: y_int
            double precision, intent(out) :: ut1_int
        end subroutine interp
        subroutine lagint(x,y,n,xint,yout) ! in :interp:interp.f
            real*8 dimension(n), intent(in) :: x
            real*8 dimension(n), intent(in) :: y
            integer, intent(hide), depend(x) :: n=shape(x, 0)
            real*8, intent(in) :: xint
            real*8, intent(out) :: yout
        end subroutine lagint
        subroutine pmut1_oceans(rjd,cor_x,cor_y,cor_ut1,cor_lod) ! in :interp:interp.f
            real*8, intent(in) :: rjd
            real*8, intent(out) :: cor_x
            real*8, intent(out) :: cor_y
            real*8, intent(out) :: cor_ut1
            real*8, intent(out) :: cor_lod
        end subroutine pmut1_oceans
        subroutine pm_gravi(rjd,cor_x,cor_y) ! in :interp:interp.f
            double precision, intent(in) :: rjd
            double precision, intent(out) :: cor_x
            double precision, intent(out) :: cor_y
        end subroutine pm_gravi
    end interface
end python module interp

! This file was auto-generated with f2py (version:2.5.1).
! and then edited by MHvK
! See:
! https://web.archive.org/web/20140822061353/http://cens.ioc.ee/projects/f2py2e
```

Compilation to a module is then with,
```
f2py -c --f77flags="-ffixed-line-length-0" interp.f interp.pyf
```

Then, as a trial,
```
from interp_mod import interp
from astropy.time import Time
from astropy.utils.iers import IERS_Auto
iers = IERS_Auto.open()

t = Time("2025-04-08T09:10:11")
iers.pm_xy(t)
# [<Quantity 0.06410133 arcsec>, <Quantity 0.38834011 arcsec>]
iers.ut1_utc(t)
# <Quantity 0.03676799 s>

interp(iers["MJD"], iers["PM_x"], iers["PM_y"], iers["UT1_UTC"], t.mjd)
(0.06414316493735002, 0.38862301914940645, 0.03677844300346548)
```

MaanasArora commented:
@mhvk - no worries, thank you for looking at this, and the helpful pointers! (Sorry, this comment ended up long.)

About exposing the feature: yes, agreed, providing the interpolation method in configuration seems much better than passing arguments. I added a configuration parameter to the IERS configuration class. It accepts `linear` (old scheme) and `tides` (with lagrange interpolation and tidal corrections).

I was able to compile `interp.f` with f2py as well using your steps and get results (more below). But while debugging an issue during compilation, not sure how, I noticed my implementation was actually very garbled! It seemed to mainly just have the `arg` such as `chi`, `omega` right, but the algorithm following was incorrectly mixed between routines and had coefficient mistakes. So, I had to go back and rewrite the whole thing before testing:

1. As part of this I converted the correct coefficient matrix (the relevant one was very large) into a CSV file, mainly by replacing the special characters (`&`, extra whitespace, and slashes), and adding a header. Then I had a small Python script write it as NumPy code, below for future reference. I'm wondering if having such a large array in the code is acceptable, though? I'm happy to use a CSV, `npy`, or another data file directly, I'm not sure what the preference is. 

<details>

<summary>
Python script to convert cleaned CSV to NumPy code
</summary>

```python
import pandas as pd
import numpy as np

csv = pd.read_csv("interp_pmut1_oceans.csv", index_col=False)
arg = csv.iloc[:, :6].to_numpy()
val = csv.iloc[:, 6:].to_numpy()

# format as Python code to create a numpy array
print("OCEANS_ARG = np.array([")
for row in arg:
    print("    [", end="")
    print(", ".join(f"{x}" for x in row), end="")
    print("],")
print("])")

print("OCEANS_VAL = np.array([")
for row in val:
    print("    [", end="")
    print(", ".join(f"{x}" for x in row), end="")
    print("],")
print("])")
```

</details>

2. Since it seems that the `PM_GRAVI` routine only impacts the x/y columns and not UT1-UTC, I have excluded that part of `interp.f` from this PR. This does mean that I only use the `tides` method for UT1-UTC currently, even if it is selected, but I think the code can be expanded quite easily as the foundation is there. We would only need to convert that routine and add it to the pipeline if the other columns are involved.

On testing, and whether just a `lagrange` option is worth it: I used this script to compare each algorithm with the f2py reference:

<details>

<summary>
Testing script
</summary>

```python
from interp_mod import interp
from astropy.time import Time
from astropy.utils import iers as iers_module
from astropy.utils.iers import IERS_Auto
import numpy as np


def get_expected_values(iers, ts):
    return np.array(
        [
            interp(iers["MJD"], iers["PM_x"], iers["PM_y"], iers["UT1_UTC"], t.mjd)
            for t in ts
        ]
    ).transpose()


def get_values_from_iers(iers, ts, method):
    iers_module.conf.ut1_utc_interpolation = method
    x, y = iers.pm_xy(ts)
    ut1_utc = iers.ut1_utc(ts)
    return np.array([x, y, ut1_utc])


def compare_interp_methods(iers, ts):
    ts_expected = get_expected_values(iers, ts)[2]
    ts_linear = get_values_from_iers(iers, ts, "linear")[2]
    ts_tides = get_values_from_iers(iers, ts, "tides")[2]

    error_linear = np.abs(ts_linear - ts_expected)
    error_tides = np.abs(ts_tides - ts_expected)

    return error_linear, error_tides


if __name__ == "__main__":
    iers = IERS_Auto.open()

    # sample 1000 points between 2024 and 2026
    start_mjd = Time("2024-01-01").mjd
    end_mjd = Time("2026-01-01").mjd
    ts = Time(np.linspace(start_mjd, end_mjd, 1000), format="mjd")

    error_linear, error_tides = compare_interp_methods(iers, ts)

    print(f"Mean absolute error (linear): {np.mean(error_linear):.3e} seconds")
    print(f"Mean absolute error (tides): {np.mean(error_tides):.3e} seconds")

    print(f"Max absolute error (linear): {np.max(error_linear):.3e} seconds")
    print(f"Max absolute error (tides): {np.max(error_tides):.3e} seconds")
```

</details>

With the current state of the branch, the results are very close to the reference:

```
Mean absolute error (linear): 2.510e-05 seconds
Mean absolute error (tides): 6.638e-13 seconds

Max absolute error (linear): 1.093e-04 seconds
Max absolute error (tides): 2.094e-12 seconds
```

The gains without tidal corrections seem much more marginal. So, I suppose it is not really worth it, so I'm not exposing a `lagrange` option, nor an `order` parameter:

```
Mean absolute error (linear): 2.510e-05 seconds
Mean absolute error (tides): 2.285e-05 seconds

Max absolute error (linear): 1.093e-04 seconds
Max absolute error (tides): 8.199e-05 seconds
```

mhvk commented:
Thanks, that is great! So, it looks like we just have two config options. 

I agree with your hesitance about large parameter arrays inside the code. Since you already made the .csv, I think we should just include that and load it on demand. After all, we are already loading the (far larger) IERS tables on demand as well, so it is not like this will meaningfully affect performance.

However, I think we should use the `Table` machinery (or `np.loadtxt`) to load the .csv; we do not want to add a dependence on pandas. Alternatively, we can include the `interp.f` file and extract the arguments programmatically from that. A bit of playing around shows this works:
```
from astropy.table import Table
from astropy.utils.data import get_pkg_data_contents

interp = get_pkg_data_contents("src/interp.f")
pre, _, gravi = interp.rpartition("data(")
oceans = pre.partition("data(")[2]

def parse_data_block(data):
    # Get header into a single line
    hdr, _, post  = data.partition(")/\n")
    # remove bits that are irrelevant
    hdr = hdr.replace("     &", "").replace("),j=1,nlines", "").strip()
    hdr = hdr.replace("(j,", "").replace("(j", "").replace("),", ",").replace("\n", "")
    # Get value lines
    values = post.partition("/\n")[0].replace("     &", "").replace(",\n", "\n")
    return Table.read("\n".join([hdr, values]), format="ascii.csv")

ocean = parse_data_block(oceans)
gravi = parse_data_block(gravi)
```

mhvk commented:
p.s. Should have added: since getting it from `interp.f` works, to me it makes most sense to just include that file. That also means that *if* a fortran compiler is available, we can easily re-check numbers.

MaanasArora commented:
Thanks, yes I just included the `interp.f` file in a `src/` directory in the module (I thought of `data/`, but this is moreso code). The snippet you shared worked nearly as-is. Yes, it is good to have the source for verification in the future, and this also means we can easily add `gravi` when need be! I'd mark this as ready for review now, except codecov doesn't seem to recognize testing coverage for the new config option.

MaanasArora commented:
Thanks @mhvk! `np.newaxis` used, makes sense. I also added the location procedure. The code was shared mostly so it didn't take up too many more lines. That said, I think I might have found a mistake in the `ARG` preparation, not entirely sure why `ARG(3)` (lp) is different in the oceans versus gravi procedure:

PMUT1_OCEANS:
```fortran
      ARG(3) = -0.00001149d0*T**4 + 0.000136d0*T**3
     .  -  0.5532d0*T**2
     .  + 129596581.0481d0*T + 1287104.79305d0
```

PM_GRAVI:
```fortran
      ARG(3) = -0.00001149d0*T**4 - 0.000136d0*T**3
     .  -  0.5532d0*T**2
     .  + 129596581.0481d0*T + 1287104.79305d0
```

The sign of the second term (`0.000136d0*T**3`) seems flipped. I've not inherited that, but happy to go either direction if you know what should be done here. To be fair even it's an error, might not be very harmful to be consistent if we're unsure. And yes, we can move to a better configuration in `ScienceState` whichever way you like!

mhvk commented:
That's a nice catch!  Looking at https://hpiers.obspm.fr/eop-pc/models/models.html_nov08, I'm led to https://hpiers.obspm.fr/eop-pc/models/PM/PM_oceanic_hf_tab.html and via `interp.f` I infer that `ARG(3)` is lp, which presumably means l', for which the bottom of the web page lists,
```
Mean anomaly of the Sun  : 
   l'= 357°.529 109 18 +   129 596 581.0481" t - 0".553 2 t2 -  0".000 136 t3 - 0".000 011 49 t4
```
So, the term should be negative in `OCEANS`, while `GRAVI` is correct. Indeed, that can also be inferred from the derivative right below in `OCEANS`.

I think that more or less settles it, but let me send an e-mail to point out the inconsistency. That also gives me a nice opportunity to verify the procedure, and ask whether there are some reference numbers we can compare with.

MaanasArora commented:
Thanks @mhvk for the email, that is likely correct indeed, so I've updated the sign to be negative!
